cmd/ttn-lw-cli/commands/end_device_templates.go

```diff
@@ -0,0 +1,436 @@
+		Use:               "list-formats",
+		Short:             "List available end device template formats",
+		PersistentPreRunE: preRun(),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			dtc, err := api.Dial(ctx, config.DeviceTemplateConverterGRPCAddress)
+			if err != nil {
+				return err
+			}
+			res, err := ttnpb.NewEndDeviceTemplateConverterClient(dtc).ListFormats(ctx, ttnpb.Empty)
+			if err != nil {
+				return err
+			}
+			return io.Write(os.Stdout, config.OutputFormat, res)
+		},
+	}
+	endDeviceTemplatesConvertCommand = &cobra.Command{
+		Use:               "convert [format-id]",
+		Short:             "Convert data to an end device template",
+		PersistentPreRunE: preRun(),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			formatID := getTemplateFormatID(cmd.Flags(), args)
+			if formatID == "" {
+				return errNoTemplateFormatID
+			}
+			data, err := getDataBytes("", cmd.Flags())
+			if err != nil {
+				return err
+			}
+
+			dtc, err := api.Dial(ctx, config.DeviceTemplateConverterGRPCAddress)
+			if err != nil {
+				return err
+			}
+			stream, err := ttnpb.NewEndDeviceTemplateConverterClient(dtc).Convert(ctx, &ttnpb.ConvertEndDeviceTemplateRequest{
+				FormatID: formatID,
+				Data:     data,
+			})
+			if err != nil {
+				return err
+			}
+
+			for {
+				dev, err := stream.Recv()
+				if err == stdio.EOF {
+					return nil
+				}
+				if err != nil {
+					return err
+				}
+				if err := io.Write(os.Stdout, config.OutputFormat, dev); err != nil {
+					return err
+				}
+			}
+		},
+	}
+	endDeviceTemplatesMapCommand = &cobra.Command{
+		Use:   "map [flags]",
+		Short: "Map end device templates",
+		Long: `Map end device templates
+
+This command matches the input template with the mapping file to create new
+templates. The mapping file contains end device templates in JSON format.
+
+The matching from input to a mapping template is, in order, by mapping key, end
+device identifiers and DevEUI. If you don't specify a mapping key, end device
+identifiers nor DevEUI, the mapping entry always matches. This is useful for
+mapping many end device templates with a generic template.
+
+Typical use cases are:
+
+1. Assigning identifiers from a mapping file to device templates matching on
+   mapping key.
+2. Mapping a device profile (i.e. MAC and PHY versions, frequency plan and class
+   B/C support) from a mapping file to many end device templates.
+
+Use the create command to create a mapping file and (optionally) the assign-euis
+command to assign EUIs to map to end device templates.`,
+		PersistentPreRunE: preRun(),
+		RunE: asBulk(func(cmd *cobra.Command, args []string) error {
+			var input ttnpb.EndDeviceTemplate
+			if inputDecoder != nil {
+				_, err := inputDecoder.Decode(&input)
+				if err != nil {
+					return err
+				}
+			} else {
+				data, err := getDataBytes("input", cmd.Flags())
+				if err != nil {
+					return err
+				}
+				if err := json.Unmarshal(data, &input); err != nil {
```

Comment:
Need to use jsonpb here, or better:


```go
reader, err := getDataReader("input", cmd.Flags())
if err != nil {
	return err
}

switch config.InputFormat {
case "json":
	inputDecoder = io.NewJSONDecoder(reader)
default:
	return fmt.Errorf("unknown input format: %s", config.InputFormat)
}
```

We may want to move that second part to a `func getInputDecoder(stdio.Reader) io.Decoder` in `utils.go` for DRY reasons, and also use that func for the main `inputDecoder`, and below.